// @tolgee/ui/lib/KeyContextMenu/KeyContextMenu.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import React from 'react';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import { styled } from '@mui/material/styles';
import { DEVTOOLS_ID, DEVTOOLS_Z_INDEX } from '../constants';
import { ThemeProvider } from '../ThemeProvider';
var ScMenuItem = styled(MenuItem)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: column;\n  height: 50px;\n  justify-content: center;\n  align-items: flex-start;\n"], ["\n  display: flex;\n  flex-direction: column;\n  height: 50px;\n  justify-content: center;\n  align-items: flex-start;\n"])));
var ScTranslation = styled('div')(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: flex;\n  padding: 3px;\n"], ["\n  display: flex;\n  padding: 3px;\n"])));
var ScKey = styled('div')(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  margin-top: 10px;\n  padding: 3px;\n  font-weight: bold;\n  font-size: 12px;\n  font-family: Monospace, 'Courier New', Courier;\n"], ["\n  display: flex;\n  margin-top: 10px;\n  padding: 3px;\n  font-weight: bold;\n  font-size: 12px;\n  font-family: Monospace, 'Courier New', Courier;\n"])));
var KeyContextMenu = /** @class */ (function (_super) {
    __extends(KeyContextMenu, _super);
    function KeyContextMenu(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            opened: false,
        };
        _this.keyDown = function (e) {
            if (e.key === 'Escape') {
                _this.setState(function (s) { return (__assign(__assign({}, s), { opened: false })); });
                _this.state.onSelect && _this.state.onSelect(null);
            }
        };
        return _this;
    }
    KeyContextMenu.prototype.show = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.setState(__assign(__assign({}, params), { opened: true }));
                return [2 /*return*/];
            });
        });
    };
    KeyContextMenu.prototype.componentDidMount = function () {
        document.addEventListener('keydown', this.keyDown);
    };
    KeyContextMenu.prototype.componentWillUnmount = function () {
        document.removeEventListener('keydown', this.keyDown);
    };
    KeyContextMenu.prototype.render = function () {
        var _this = this;
        return (React.createElement(ThemeProvider, null, this.state.opened && (React.createElement(Menu, { disablePortal: true, disableEnforceFocus: true, anchorEl: this.state.openEvent.target, anchorOrigin: {
                vertical: 'bottom',
                horizontal: 'center',
            }, open: true, onClose: function () {
                _this.setState({ opened: false });
                _this.state.onSelect(null);
            }, container: document.getElementById(DEVTOOLS_ID), style: { zIndex: DEVTOOLS_Z_INDEX } }, Array.from(this.state.keys).map(function (key, index) { return (React.createElement(ScMenuItem, { onClick: function () {
                _this.state.onSelect(key);
                setTimeout(function () {
                    _this.setState({ opened: false });
                });
            }, key: index },
            React.createElement(ScTranslation, { "data-cy": "context-menu-value" }, _this.props.dependencies.translationService.getFromCacheOrFallback(key)),
            React.createElement(ScKey, { "data-cy": "context-menu-key" }, key))); })))));
    };
    return KeyContextMenu;
}(React.Component));
export { KeyContextMenu };
var templateObject_1, templateObject_2, templateObject_3;
