// @tolgee/ui/lib/KeyDialog/KeyForm.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import { RESTRICTED_ASCENDANT_ATTRIBUTE } from '@tolgee/core';
import { styled } from '@mui/material/styles';
import IconButton from '@mui/material/IconButton';
import Button from '@mui/material/Button';
import OpenInNew from '@mui/icons-material/OpenInNew';
import { TextHelper } from '@tolgee/core';
import { TranslationFields } from './TranslationFields';
import { LanguageSelect } from './LanguageSelect';
import { LoadingButton } from '../common/LoadingButton';
import { ScreenshotGallery } from './ScreenshotGallery/ScreenshotGallery';
import { ScFieldTitle } from '../common/FieldTitle';
import { useDialogContext, useDialogDispatch, } from './TranslationDialogContextProvider';
var ScContainer = styled('div')(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  font-family: Rubik, Roboto, Arial;\n  padding: 20px;\n  box-sizing: border-box;\n  max-width: 100%;\n  width: 700px;\n  display: flex;\n  flex-direction: column;\n"], ["\n  font-family: Rubik, Roboto, Arial;\n  padding: 20px;\n  box-sizing: border-box;\n  max-width: 100%;\n  width: 700px;\n  display: flex;\n  flex-direction: column;\n"])));
var ScHeading = styled('div')(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  gap: 5px;\n"], ["\n  display: flex;\n  align-items: center;\n  gap: 5px;\n"])));
var ScHeadingTitle = styled('div')(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  margin: 0px;\n  font-size: 19px;\n"], ["\n  display: flex;\n  margin: 0px;\n  font-size: 19px;\n"])));
var ScHeadingRight = styled('div')(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  display: flex;\n  justify-content: flex-end;\n  align-items: center;\n  flex-grow: 1;\n"], ["\n  display: flex;\n  justify-content: flex-end;\n  align-items: center;\n  flex-grow: 1;\n"])));
var ScKey = styled('p')(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  margin: 0px;\n"], ["\n  margin: 0px;\n"])));
var ScKeyHint = styled('span')(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  color: grey;\n"], ["\n  color: grey;\n"])));
var ScFieldsWrapper = styled('div')(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  margin-top: 20px;\n"], ["\n  margin-top: 20px;\n"])));
var ScGalleryWrapper = styled('div')(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  margin-top: 10px;\n"], ["\n  margin-top: 10px;\n"])));
var ScControls = styled('div')(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  display: flex;\n  justify-content: flex-end;\n  margin-top: 20px;\n  min-height: 36px;\n"], ["\n  display: flex;\n  justify-content: flex-end;\n  margin-top: 20px;\n  min-height: 36px;\n"])));
var ScRestriction = styled('div')(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  margin-top: 8px;\n  color: ", ";\n"], ["\n  margin-top: 8px;\n  color: ", ";\n"])), function (_a) {
    var theme = _a.theme;
    return theme.palette.text.secondary;
});
var ScError = styled('div')(templateObject_11 || (templateObject_11 = __makeTemplateObject(["\n  color: red;\n"], ["\n  color: red;\n"])));
export var KeyForm = function () {
    var _a;
    var dispatch = useDialogDispatch();
    var linkToPlatform = useDialogContext(function (c) { return c.linkToPlatform; });
    var useBrowserWindow = useDialogContext(function (c) { return c.useBrowserWindow; });
    var dependencies = useDialogContext(function (c) { return c.dependencies; });
    var input = useDialogContext(function (c) { return c.input; });
    var translations = useDialogContext(function (c) { return c.translations; });
    var formDisabled = useDialogContext(function (c) { return c.formDisabled; });
    var loading = useDialogContext(function (c) { return c.loading; });
    var error = useDialogContext(function (c) { return c.error; });
    var saving = useDialogContext(function (c) { return c.saving; });
    var success = useDialogContext(function (c) { return c.success; });
    var screenshotsView = dependencies.coreService.isAuthorizedTo('screenshots.view');
    var setUseBrowserWindow = function (value) {
        dispatch({ type: 'SET_USE_BROWSER_WINDOW', payload: value });
    };
    var onClose = function () {
        dispatch({ type: 'ON_CLOSE' });
    };
    var onSave = function () {
        dispatch({ type: 'ON_SAVE' });
    };
    return (React.createElement(ScContainer, __assign({}, (_a = {}, _a[RESTRICTED_ASCENDANT_ATTRIBUTE] = 'true', _a)),
        React.createElement(ScHeading, null,
            React.createElement("a", { href: linkToPlatform, target: "_blank", rel: "noreferrer noopener", id: "_tolgee_platform_link" },
                React.createElement("svg", { viewBox: "0 0 200 200", xmlns: "http://www.w3.org/2000/svg", opacity: "0.99", fill: "#822B55", style: {
                        fillRule: 'evenodd',
                        clipRule: 'evenodd',
                        strokeLinejoin: 'round',
                        strokeMiterlimit: 2,
                        height: 23,
                    } },
                    React.createElement("path", { d: "M97.16,7.27a16.94,16.94,0,0,0-1.9,24.47,16.36,16.36,0,0,0,5,3.83,3.23,3.23,0,0,1-2.9,5.77,23.14,23.14,0,0,1-11.41-13C73.83,31.1,63.46,37.09,52.82,46.51c-27.44,24.3-34.35,61.74-16.38,85.26-4.57,5.79-8,12.22-8.9,18.69a20.88,20.88,0,0,0,5.62,18c9.18,9.61,21.42,7.13,31.26,5.14,6.58-1.34,12.8-2.6,16.5-.23,3.22,2.07,3.47,3.87,3.61,4.45,2.1,9.32-5.79,13.89-7.67,16.27a1.48,1.48,0,0,0,1.13,2.4c3.48,0,9-1.18,12.34-4.08s7.16-7.9,5.89-16.32c-.08-.5-.18-1-.32-1.58-.86-3.35-3.1-7.57-8.61-11.09-7.72-4.95-17-3.07-25.22-1.41-9.76,2-16,2.85-20.37-1.71a9.13,9.13,0,0,1-2.46-8.19c.54-3.77,2.65-7.89,5.62-11.86,21.71,16.89,56.87,13.47,82.67-9.39a75.34,75.34,0,0,0,20.81-28.09A23.14,23.14,0,0,1,134.8,89a3.23,3.23,0,0,1,6.08-2.19,16.37,16.37,0,0,0,3.2,5.39,16.85,16.85,0,1,0,11.48-28,3.23,3.23,0,0,1-.51-6.44,23.41,23.41,0,0,1,12.88,2.69c2.6-14.08,3.34-31.41-2.06-37.51-4.08-4.61-20.62-8-35.18-7.76A23.48,23.48,0,0,1,130.8,25a3.23,3.23,0,0,1-6.33-1.28A16.94,16.94,0,0,0,97.16,7.27Zm63.25,21a5.29,5.29,0,0,1-.57,6.19c-1.29,1.14-2.72-.51-4.1-2.06s-3.1-3.42-1.81-4.56A5.74,5.74,0,0,1,160.41,28.27Z" }))),
            React.createElement(ScHeadingTitle, null, "Quick translation"),
            !useBrowserWindow && (React.createElement(IconButton, { title: "Open in new window", onClick: function () { return setUseBrowserWindow(true); }, color: "inherit", size: "small" },
                React.createElement(OpenInNew, { fontSize: "small" }))),
            React.createElement(ScHeadingRight, null,
                React.createElement(LanguageSelect, null))),
        React.createElement(ScFieldTitle, null, "Key"),
        React.createElement(ScKey, null,
            input && TextHelper.removeEscapes(input),
            React.createElement(ScKeyHint, null, (translations === null || translations === void 0 ? void 0 : translations.keyId) === undefined && " (key doesn't exist yet)")),
        React.createElement(ScFieldsWrapper, null,
            React.createElement(TranslationFields, null)),
        screenshotsView && (React.createElement(ScGalleryWrapper, null,
            React.createElement(ScreenshotGallery, null))),
        formDisabled && !loading && (React.createElement(ScRestriction, null, "Modification is restricted due to missing ".concat((translations === null || translations === void 0 ? void 0 : translations.keyId) !== undefined ? 'translations.edit' : 'keys.edit', " scope in current api key settings."))),
        error && React.createElement(ScError, null, error),
        React.createElement(ScControls, null,
            React.createElement(Button, { onClick: onClose, color: "secondary" }, useBrowserWindow ? 'Close' : 'Cancel'),
            React.createElement(LoadingButton, { loading: saving, disabled: saving || formDisabled, onClick: onSave, color: "primary", variant: "contained", style: { marginLeft: '10px' } }, success
                ? 'Saved! ✓'
                : (translations === null || translations === void 0 ? void 0 : translations.keyId) === undefined
                    ? 'Create'
                    : 'Update'))));
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11;
